More robust CLI parameter handling

ozmorph opened the pull request:
This merge request aims to provide a basis for solving some of the tasks in #74, specifically `Ensure for command line options taking an input file name we check the file exists`.

I switched out the series of if/else statements for checking CLI parameters to use switch case statements for legibility, performance, and consistent type checking. I added a `ParseArg()` function to standardize the use of `sscanf()` calls in case a 3rd party C++ library is eventually used like Boost which has a much more robust string parsing functionality than the C library.

I've also built two "usage" strings that will displayed to the user either in error (bad argument or no argument) or when the `/H` argument is specified. It builds on the documentation work from #164 by providing a more detailed description of required and optional parameters for the `CovidSim.exe` program.

I've also began to replace usage of `fprintf()` with `iostream` for the simple purpose of using C++ idioms since this project will eventually use more and more C++ as time goes on.

Please let me know if there are any changes you would like me to make!

ozmorph commented:
If this merge request is accepted. I will open a subsequent merge request to refactor all CLI specific code into a `CLI.cpp` file to help de-clutter `CovidSim.cpp`.

ozmorph commented:
First off, thank you for the very detailed response!

I agree with all of your proposed changes.  While implementing this PR, I didn't want to deviate too far from the current C logic in one swipe because changes on a project of this size and importance usually need to be slow and tenable in order to be accepted.

If adding a C++ class called CmdLineArgs that implements this functionality is something the team is open to accepting, then I will gladly proceed with making these more advanced changes.

Also with regard to your design since it wasn't specified, do you want to keep the traditional DOS way of specifying arguments (`/P:param_file.txt`) or is there something else you were wanting to switch to?

Please give me a day or two to update my PR.

NeilFerguson commented:
I am happy with changes so long as they preserve backwards compatibility. So I’d prefer the command line syntax to not change.

Best,

Neil
________________________________
From: ozmorph <notifications@github.com>
Sent: Wednesday, May 13, 2020 1:33:02 PM
To: mrc-ide/covid-sim <covid-sim@noreply.github.com>
Cc: Ferguson, Neil M <neil.ferguson@imperial.ac.uk>; Review requested <review_requested@noreply.github.com>
Subject: Re: [mrc-ide/covid-sim] More robust CLI parameter handling (#228)


First off, thank you for the very detailed response!

I agree with all of your proposed changes. While implementing this PR, I didn't want to deviate too far from the current C logic in one swipe because changes on a project of this size and importance usually need to be slow and tenable in order to be accepted.

If adding a C++ class called CmdLineArgs that implements this functionality is something the team is open to accepting, then I will gladly proceed with making these more advanced changes.

Also with regard to your design since it wasn't specified, do you want to keep the traditional DOS way of specifying arguments (/P:param_file.txt) or is there something else you were wanting to switch to?

Please give me a day or two to update my PR.

—
You are receiving this because your review was requested.
Reply to this email directly, view it on GitHub<https://github.com/mrc-ide/covid-sim/pull/228#issuecomment-627951397>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AFRK7N3QET66442JI626CB3RRKHP5ANCNFSM4M67RP2Q>.

matt-gretton-dann commented:
As Neil says we should preserve the current command-line behaviour - we're re-engineering not changing the UI :-).

Also feel free to share the changes in a half-finished state so that there is some early review - and we can agree that the direction they're going in is correct (either raise a new PR & close this one or change this one into a draft PR until its complete).

ozmorph commented:
I agree wholeheartedly with keeping the UI/CLI consistent and backwards-compatible, that's why I asked.

I will push my changes (even if half-finished) whenever I get them to a state where the regression tests still pass so that you all can run the code as-is while reviewing.

ozmorph commented:
@matt-gretton-dann As requested, I have pushed changes in a test-passing but half-finished/non-mergable state. When you get the time, please review the changes I have made.

A summary of my changes:

* Added new CLI.hpp and CLI.cpp file which will eventually contain all CLI-specific code from CovidSim.cpp. (***note*** I used .hpp for the header because it includes template code)
* Added CmdLineArgs class to CLI.hpp which includes two functions `add_option()` and `parse()`
* Changed CovidSim.cpp to parse the RNG seeds and P/PP options using CmdLineArgs.
* Removed need for C-style string parsing. I replaced `sscanf()` with istringstream's `>>` operator and replaced `char*` with `std::string`.
* Stored all options in a `std::map` which maps to a function call with the output variable T being bound via `std::bind`.

The `add_option()` function requires a template parameter because I was unable to figure out a clean way to pass a type T within the `std::function` declaration for `std::map<>` and the function `add_option()`. If you (or anyone else reading this) has a solution that doesn't involve making wrapper classes to Null functions, please let me know. I hope that adding the type to a function template is clean enough for this current solution.

Let me know what you think! If you're happy with how I've set up `CmdLineArgs`, then I'll change all of the options to use this new format and push those for review over the next day or two.

ozmorph commented:
> The `add_option()` function requires a template parameter because I was unable to figure out a clean way to pass a type T within the `std::function` declaration for `std::map<>` and the function `add_option()`. If you (or anyone else reading this) has a solution that doesn't involve making wrapper classes to Null functions, please let me know. I hope that adding the type to a function template is clean enough for this current solution.

Addendum: I have figured out an alternative (and I think cleaner) way to not have to specialize the add_option function call in `CovidSim.cpp`. I've introduced `add_string_option()` and `add_integral_option()`. Integral options should use the `parse_integral` option no matter what their type. String options on the other hand need the option to use a different parse function because the string could be a file path, a directory, a plain name for a variable, an output path, or a choice in limited set of value (like the `/BM` option mentioned). Each of which require their own logic.

I'm done touching this pull request for now and will keep my commits local until I hear back from a reviewer.

Cheers!

ozmorph commented:
Under the assumption that no one is currently reviewing my changes at the moment, I've gone ahead and pushed all of my commits from today (which are still passing regression tests).

This latest batch of commits has fully converted every single CLI argument to use the CmdLineArgs class. In addition I've added several new features:

* `add_integral_option()` is now `add_number_option()` to account for the addition of a `std::is_floating_point` check for float and double parameters.
* New `add_custom_option()` accepts a `CmdLineArgs::ParserFn` for dealing with options that have multiple side effects (like `/SS` and `/BM`)
* I've changed all `char` arrays to use `std::string` to reduce the use of `sprintf()` throughout the code

I'm sure there are changes that will make this pull request better, but I consider the current state to be representative of an MVP for the new CLI argument parsing.

Please let me know if you have any questions or concerns!

ozmorph commented:
I'm pushing a rebase of this branch on top of master as of 30 minutes ago which resolves conflicts arising from the new `/NR` CLI argument from #306 and some lines from #312.

ozmorph commented:
@matt-gretton-dann Is the direction this PR is going what you had in mind? I figured one additional thing I could add is the doc string of each parameter as another argument to the `add_option()`.  Each string would be stored in a sorted map and the detailed help string could be generated via a for loop rather than manually curating a large help string as it is now.

Also as an FYI, I have begun work on a separate branch that starts to tackle the other part #74 with regard to parameter files. I'm using a similar technique to the  `parse_number()` function of this PR and parsing the parameters via the storage type rather than a void* and sscanf format pair. If you have ideas regarding the design of parameter file parsing, I would be happy to hear them!

Here is the current design of my parser class:

```c++
class ParamReader {
public:
    ParamReader(std::string const& param_file, std::string const& preparam_file, std::string const& admin_file);

    template<typename T>
    bool extract(std::string const& param, T& output, T default);

    template<typename T>
    void extract_or_exit(std::string const& param, T& output);

private:
    void parse_param_file(std::string param_file);

    std::unordered_map<std::string, std::string> m_param_value_map;
};
``` 

Thanks!

ozmorph commented:
Also, I will resolve all conflicts between this branch and master shortly.

ozmorph commented:
@CloneDeath @matt-gretton-dann @NeilFerguson

I no longer consider this pull request to be in a work-in-progress. It is ready for final review.

ozmorph commented:
Rebased with master to fix conflicts. Pending final review.

ozmorph commented:
Rebased with master and fixed conflicts, no other changes made.

weshinsley commented:
This looks good @ozmorph - sorry for all the delays - I have about 4 PRs in the queue, and then I'll be on this one.

ozmorph commented:
> This looks good @ozmorph - sorry for all the delays - I have about 4 PRs in the queue, and then I'll be on this one.

No worries! I know my PR's are large and take longer to review. I'm happy to periodically rebase and fix conflicts.

weshinsley commented:
I am keen to get this merged, as I think it looks good and cleans up that part of the code nicely - I'm just been asked to hold back as some other (small I think) tweaks are being done to the CLI params at the moment...

ozmorph commented:
Sounds good. Thanks for the update!

I'll keep any eye out and resolve conflicts when they arise.

weshinsley commented:
Sorry - I seem to have broken bitmap.cpp in the merge...! Master has Bitmap.cpp:113 as
char buf[3000], OutF[3000];

ozmorph commented:
> Sorry - I seem to have broken bitmap.cpp in the merge...! Master has Bitmap.cpp:113 as
> char buf[3000], OutF[3000];

No big deal! I've rebased on top of master with the fix.

matt-gretton-dann commented on src/CovidSim.cpp:
```diff
@@ -480,8 +361,45 @@ int main(int argc, char* argv[])
+void parse_save_snapshot_option(std::string const& input) {
+	auto sep = input.find_first_of(',');
+	if (sep == std::string::npos) {
+		std::cerr << "Argument value didn't use a ',': " << input << std::endl;
```
This message could be improved - what should the argument look like?

ozmorph replied:
New comment should be better now.

ozmorph commented on src/CLI.cpp:
```diff
@@ -0,0 +1,178 @@
+    output = input;
+}
+
+template<typename T>
```
I agree with using `std::stoul()` and `std::stod()` instead of `std::istringstream` here.

ozmorph replied:
There is not `parse_integer()` and `parse_double()` and also `add_double_option()` and `add_integer_option()`. These use `std::stod()` and `std::stoi()`

ozmorph commented on src/CLI.hpp:
```diff
@@ -0,0 +1,90 @@
+#ifndef COVIDSIM_CLI_HPP_INCLUDED_
```
Done

ozmorph commented on src/CLI.hpp:
```diff
@@ -0,0 +1,90 @@
+    int parse(int argc, char* argv[], Param& P);
+
+private:
+    std::map<std::string, ParserFn> m_option_map;
```
Done

ozmorph commented on src/CLI.cpp:
```diff
@@ -0,0 +1,200 @@
+
+template<typename T>
+void CmdLineArgs::add_number_option(std::string const&& option, T& output) {
+    if (m_option_map.find(option) != m_option_map.cend()) {
```
Done

ozmorph commented on src/CLI.cpp:
```diff
@@ -0,0 +1,200 @@
+    iss >> output;
+    if (iss.fail()) {
+        if (output == std::numeric_limits<T>::max()) {
+            std::cerr << "OVERFLOW: detected a number larger than " << std::numeric_limits<T>::max()
```
Done

ozmorph commented on src/CovidSim.cpp:
```diff
@@ -100,29 +111,26 @@ int32_t *bmInfected; // The number of infected people in each bitmap pixel.
 int PlaceDistDistrib[NUM_PLACE_TYPES][MAX_DIST], PlaceSizeDistrib[NUM_PLACE_TYPES][MAX_PLACE_SIZE];
 
 
+
```
Removed

ozmorph commented on src/SetupModel.cpp:
```diff
@@ -1684,7 +1684,7 @@ void SetupAirports(void)
 
 
-void AssignHouseholdAges(int n, int pers, int tn)
+void AssignHouseholdAges(int n, int pers, int tn, std::string const& RegDemogFile)
```
I have gone through and renamed every variable I've touched in this pull request to use snake_case. In addition, I have pulled most of the global variables I touched into local variables to help de-clutter the global namespace and make future refactoring easier.

matt-gretton-dann commented on src/CLI.cpp:
```diff
@@ -0,0 +1,197 @@
+{
+	try
+	{
+		output = std::stoi(input);
```
Can we do something like:
```suggestion
std::size_t pos;
output = std::stoi(input, &pos);
if (pos != input.len)
{
  ERR_CRITICAL_FMT("Unable to completely parse integer: %s\n", input.c_str());
}
```

Similarly for float parsing later on.

ozmorph replied:
Done

matt-gretton-dann commented on src/SetupModel.cpp:
```diff
@@ -1688,7 +1690,7 @@ void SetupAirports(void)
 
 
-void AssignHouseholdAges(int n, int pers, int tn)
+void AssignHouseholdAges(int n, int pers, int tn, std::string const& reg_demog_file)
```
This feels like the wrong interface - we only test `reg_demog_file` for emptiness maybe something like:
```suggestion
void AssignHouseholdAges(int n, int pers, int tn, bool do_adunit_demog)
```
Instead?

ozmorph replied:
Done

matt-gretton-dann commented on src/SetupModel.h:
```diff
@@ -1,15 +1,17 @@
-void SetupPopulation(char*, char*);
+#include <string>
+
+void SetupModel(std::string const&, std::string const&, std::string const&, std::string const&, std::string const&, std::string const&, std::string const&);
```
Comments on function parameters please.  Especially those that take multiple strings, and also indicate which ones are allowed to be empty, and what that means (that we don't load/save/do a calculation).

ozmorph replied:
I've documented `SetupModel.h` to the best of my understanding of what each function does. Feel free to point out inaccuracies/inconsistencies.

ozmorph commented:
> This is a large PR that has reached the stage where everytime I review it I'm going to find something to change/modify.
> 
> I have one or two final comments - but then unless I see a major issue I am going to approve as this is a general improvement and any future changes can be built on top of it.

@matt-gretton-dann I do apologize for the size and scope of this PR. I know that large refactors like these are burdensome to review, especially when they touch many files and interfaces. At the beginning, I was not confident that a C++ interface with C-style strings was the appropriate midterm solution. But now I hope that it's complete and enough of an improvement that it all warrants itself.

I would be glad to add unit tests to this PR, but I'm also fine with adding them in a follow-on PR. It may take a few days for me to add them since I'm new to GoogleTests and am short on time this week.

I'm also happy to make any additional changes you see fit, but I agree that there's plenty of future work that could be built on top after this has been merged.

Thanks for the great feedback and review!
